diff -urN aaa/gnome-session-2.4.1/gnome-session/logout.c gnome-session-2.4.1/gnome-session/logout.c
--- aaa/gnome-session-2.4.1/gnome-session/logout.c	2003-08-07 14:15:41.000000000 +0200
+++ gnome-session-2.4.1/gnome-session/logout.c	2003-11-20 13:53:19.000000000 +0100
@@ -21,6 +21,7 @@
 #include <stdio.h>
 #include <string.h>
 #include <unistd.h>
+#include <sys/stat.h>
 #include <gtk/gtk.h>
 #include <gtk/gtkinvisible.h>
 #include <gdk/gdkx.h>
@@ -334,6 +335,7 @@
   GError *error = NULL;
   GdkScreen *screen;
   int monitor;
+	struct stat st;
 
   gsm_verbose ("display_gui: showing logout dialog\n");
 
@@ -427,7 +429,9 @@
    */
   s = g_strconcat ("/var/lock/console/", g_get_user_name (), NULL);
   t = g_strconcat ("/var/run/console/", g_get_user_name (), NULL);
-  if (((geteuid () == 0) || g_file_exists (t) || g_file_exists(s)) &&
+	stat (halt_command[0], &st);
+  if (((geteuid () == 0) || ((g_file_exists (t) || g_file_exists(s)) &&
+      (st.st_uid == 0) && (st.st_mode & S_ISUID))) &&
       access (halt_command[0], X_OK) == 0)
     {
       GtkWidget *title, *spacer;
