diff -urN gnome-session-2.16.2/configure.in gnome-session-2.16.2.new/configure.in
--- gnome-session-2.16.2/configure.in	2006-11-20 15:07:45.000000000 +0100
+++ gnome-session-2.16.2.new/configure.in	2006-11-30 19:13:05.000000000 +0100
@@ -65,21 +65,6 @@
 	ESOUND_MODULE="esound >= $ESOUND_REQUIRED"
 	PKG_CHECK_MODULES(ESOUND, $ESOUND_MODULE)
 	AC_DEFINE(HAVE_ESD, 1, [Define if you have the ESound Daemon])
-
-	ESDSERVERDIR=`$PKG_CONFIG --variable=esd_serverdir esound`
-	old_path=$PATH
-	if test x"$ESDSERVERDIR" != x; then
-		PATH=$ESDSERVERDIR:$PATH
-	fi
-
-	AC_PATH_PROG(ESD_SERVER, esd, no)
-	if test x"$ESD_SERVER" = xno; then
-		AC_MSG_ERROR([esd executable not found in your path - should be installed with ESound])
-	fi
-
-	AC_SUBST(ESD_SERVER)
-	PATH=$old_path
-
 else
 	ESOUND_MODULE=""
 fi
@@ -127,13 +112,6 @@
 
 AC_SUBST(GCONF_SANITY_CHECK)
 PATH=$old_path
-	
-AC_PATH_PROG(GNOME_KEYRING_DAEMON, gnome-keyring-daemon, no)
-if test x"$GNOME_KEYRING_DAEMON" = xno; then
-  AC_MSG_ERROR([gnome-keyring-daemon executable not found in your path - should be installed with gnome-keyring])
-fi
-	
-AC_SUBST(GNOME_KEYRING_DAEMON)
 
 dnl X development libraries check
 
