diff -aurN gnome-session-2.4.2.orig/gnome-session/gnome-wm gnome-session-2.4.2/gnome-session/gnome-wm
--- gnome-session-2.4.2.orig/gnome-session/gnome-wm	2002-08-06 00:13:33.000000000 +0200
+++ gnome-session-2.4.2/gnome-session/gnome-wm	2004-02-06 20:36:58.529105760 +0100
@@ -43,7 +43,7 @@
 IFS=":"
 if [ -z "$WINDOW_MANAGER" ] ; then
   if [ -z "$DEFWM" -o "x$DEFWM" = "xgnome-wm" ]; then
-    for wm in metacity sawfish sawmill enlightenment icewm wmaker fvwm2 qvwm fvwm twm kwm ; do
+    for wm in metacity sawfish sawmill openbox enlightenment icewm wmaker fvwm2 qvwm fvwm twm kwm ; do
       for dir in $PATH ; do
 	if [ -x "$dir/$wm" ] ; then
 	  WINDOW_MANAGER="$dir/$wm"
@@ -72,6 +72,10 @@
     sawfish|sawmill|metacity)
       OPT1=--sm-client-id=$SMID
       ;;
+    openbox)
+      OPT1=--sm-client-id
+      OPT2=$SMID
+      ;;
     enlightenment|twm)
       OPT1=-clientId
       OPT2=$SMID
