--- gnome-session-2.22.0/configure.in~	2008-03-10 21:44:33.000000000 +0100
+++ gnome-session-2.22.0/configure.in	2008-03-12 13:22:44.243241843 +0100
@@ -107,7 +107,7 @@
 	
 AC_PATH_PROG(GNOME_KEYRING_DAEMON, gnome-keyring-daemon, no)
 if test x"$GNOME_KEYRING_DAEMON" = xno; then
-  AC_MSG_ERROR([gnome-keyring-daemon executable not found in your path - should be installed with gnome-keyring])
+	GNOME_KEYRING_DAEMON="/usr/bin/gnome-keyring-daemon"
 fi
 	
 AC_SUBST(GNOME_KEYRING_DAEMON)
