diff -aurN gnome-session-2.15.1.orig/gnome-session/main.c gnome-session-2.15.1/gnome-session/main.c
--- gnome-session-2.15.1.orig/gnome-session/main.c	2006-04-24 23:48:26.000000000 +0200
+++ gnome-session-2.15.1/gnome-session/main.c	2006-06-15 13:43:59.567164120 +0200
@@ -338,6 +338,7 @@
   if (getenv ("GSM_VERBOSE_DEBUG"))
     gsm_set_verbose (TRUE);
 
+#if 0
   /* Help eradicate the critical warnings in unstable releases of GNOME */
   versions = g_strsplit (VERSION, ".", 3);
   if (versions && versions [0] && versions [1])
@@ -351,7 +352,8 @@
 	}
     }
   g_strfreev (versions);
-      
+#endif
+ 
   set_lang();
   set_gtk1_theme_rcfile ();
 
