diff -Nuard gnome-session-2.4.1.orig/gnome-session/main.c gnome-session-2.4.1/gnome-session/main.c
--- gnome-session-2.4.1.orig/gnome-session/main.c	2003-10-14 12:47:52.000000000 +0200
+++ gnome-session-2.4.1/gnome-session/main.c	2003-11-13 22:48:12.000000000 +0100
@@ -419,9 +419,29 @@
   n_screens = gdk_display_get_n_screens (display);
   for (i = 0; i < n_screens; i++)
     {
+      gboolean calculate_dpi;
+
       screen = gdk_display_get_screen (display, i);
       root_window = gdk_screen_get_root_window (screen);
 
+      /* try to guess right DPI */
+      calculate_dpi = gconf_client_get_bool (client, "/desktop/gnome/font_rendering/calculate_dpi" ,NULL);
+      
+      if (calculate_dpi == TRUE)
+        {
+          int width_mm, width_px;
+          double xres;
+	   int dpi;
+	  
+          width_mm = gdk_screen_get_width_mm(screen);
+          width_px  = gdk_screen_get_width(screen);
+
+          xres = ((((double)width_px) * 25.4) / ((double)width_mm));
+          dpi = (int)(xres + 0.5);
+	  
+          gconf_client_set_float (client, "/desktop/gnome/font_rendering/dpi", dpi ,NULL);
+        }
+
       if (get_resolution (client, i, keys, &width, &height))
 	{
 	  XRRScreenSize *sizes;
