--- gnome-session-2.24.1/data/gnome-session.schemas.in.in~	2008-10-22 11:15:38.000000000 +0200
+++ gnome-session-2.24.1/data/gnome-session.schemas.in.in	2008-10-23 23:50:41.627830429 +0200
@@ -7,7 +7,7 @@
          <applyto>/apps/gnome-session/options/show_splash_screen</applyto>
          <owner>gnome</owner>
          <type>bool</type>
-         <default>false</default>
+         <default>true</default>
          <locale name="C">
             <short>Show the splash screen</short>
             <long>Show the splash screen when the session starts up</long>
