diff -urN gnome-session-2.17.92/data/session-properties.desktop.in.in gnome-session-2.17.92.new/data/session-properties.desktop.in.in
--- gnome-session-2.17.92/data/session-properties.desktop.in.in	2007-02-26 21:39:19.000000000 +0100
+++ gnome-session-2.17.92.new/data/session-properties.desktop.in.in	2007-03-04 13:55:21.000000000 +0100
@@ -7,7 +7,7 @@
 Terminal=false
 Type=Application
 StartupNotify=true
-Categories=Settings;X-GNOME-PersonalSettings;
+Categories=GTK;GNOME;Settings;X-GNOME-PersonalSettings;
 OnlyShowIn=GNOME;
 X-GNOME-Bugzilla-Bugzilla=GNOME
 X-GNOME-Bugzilla-Product=gnome-session
